File: Server/src/main/content/region/desert/handlers/DesertZone.java
package content.region.desert.handlers;

import core.cache.def.impl.ItemDefinition;
import core.game.node.entity.Entity;
import core.game.node.entity.combat.ImpactHandler.HitsplatType;
import core.game.node.entity.player.Player;
import core.game.node.item.Item;
import core.game.system.task.Pulse;
import core.game.world.GameWorld;
import core.game.world.map.zone.MapZone;
import core.game.world.map.zone.ZoneBorders;
import core.game.world.map.zone.ZoneBuilder;
import core.game.world.update.flag.context.Animation;
import core.plugin.Initializable;
import core.plugin.Plugin;
import core.utilities.RandomFunction;

import java.util.ArrayList;
import java.util.List;

@Initializable
public final class DesertZone extends MapZone implements Plugin<Object> {

    private static final Item[] WATER_SKINS = new Item[]{new Item(1823), new Item(1825), new Item(1827), new Item(1829)};

    private static final int[][] VESSILS = new int[][]{{1937, 1935}, {1929, 1925}, {1921, 1923}, {227, 229}};

    private static final Animation ANIMATION = new Animation(829);

    private static final List<Player> PLAYERS = new ArrayList<>(20);

    private static final Pulse pulse = new Pulse(3) {
        @Override
        public boolean pulse() {
            for (Player player : PLAYERS) {
                if (!player.getAttribute("tutorial:complete", false) || player.getInterfaceManager().isOpened() || player.getInterfaceManager().hasChatbox() || player.getLocks().isMovementLocked()) {
                    continue;
                }
                if (player.getAttribute("desert-delay", -1) < GameWorld.getTicks()) {
                    effect(player);
                }
            }
            return PLAYERS.isEmpty();
        }
    };

    private static void effect(Player p) {
        p.setAttribute("desert-delay", GameWorld.getTicks() + getDelay(p));
        evaporate(p);
        if (drink(p)) {
            return;
        }
        p.getImpactHandler().manualHit(p, RandomFunction.random(1, p.getLocation().getY() < 2990 ? 12 : 8), HitsplatType.NORMAL);
        p.getPacketDispatch().sendMessage("You start dying of thirst while you're in the desert.");
    }


    /**
     * Evaporate.
     *
     * @param p the p
     */
    public static void evaporate(Player p) {
        for (int i = 0; i < VESSILS.length; i++) {
            if (p.getInventory().contains(VESSILS[i][0], 1)) {
                if (p.getInventory().remove(new Item(VESSILS[i][0]))) {
                    p.getInventory().add(new Item(VESSILS[i][1]));
                    p.getPacketDispatch().sendMessage("The water in your " + ItemDefinition.forId(VESSILS[i][0]).getName().toLowerCase().replace("of water", "").trim() + " evaporates in the desert heat.");
                }
            }
        }
    }


    /**
     * Drink boolean.
     *
     * @param p the p
     * @return the boolean
     */
    public static boolean drink(Player p) {
        for (Item i : WATER_SKINS) {
            if (p.getInventory().containsItem(i) && p.getInventory().remove(i)) {
                p.getInventory().add(new Item(i.getId() + 2));
                p.animate(ANIMATION);
                p.getPacketDispatch().sendMessage("You take a drink of water.");
                return true;
            }
        }
        if (p.getInventory().contains(1831, 1)) {
            p.getPacketDispatch().sendMessage("Perhaps you should fill up one of your empty waterskins.");
        } else {
            p.getPacketDispatch().sendMessage("You should get a waterskin for any travelling in the desert.");
        }
        return false;
    }

    private static int getDelay(Player player) {
        int delay = 116;
        if (player.getEquipment().contains(1833, 1)) {
            delay += 17;
        }
        if (player.getEquipment().contains(1835, 1)) {
            delay += 17;
        }
        if (player.getEquipment().contains(1837, 1)) {
            delay += 17;
        }
        return delay;
    }


    /**
     * Instantiates a new Desert zone.
     */
    public DesertZone() {
        super("Desert Zone", true);
    }

    @Override
    public Plugin<Object> newInstance(Object arg) throws Throwable {
        ZoneBuilder.configure(this);
        return this;
    }

    @Override
    public Object fireEvent(String identifier, Object... args) {
        return null;
    }

    @Override
    public void configure() {
        ZoneBorders borders = new ZoneBorders(3063, 2725, 3544, 3115);
        borders.addException(new ZoneBorders(3152, 2961, 3191, 2999));// Bandit
        // camp
        borders.addException(new ZoneBorders(3398, 2914, 3450, 2941));//Nardah
        borders.addException(new ZoneBorders(3410, 2883, 3450, 2941));//Nardah2
        borders.addException(new ZoneBorders(3147, 3019, 3185, 3059));// Bedabin
        borders.addException(new ZoneBorders(3217, 2881, 3248, 2914));// pyramid
        borders.addException(new ZoneBorders(3264, 2752, 3323, 2810));// Sophanem
        borders.addException(new ZoneBorders(3326, 2942, 3371, 3006)); //Pollnivneach
        borders.addException(new ZoneBorders(3007, 4672, 3071, 4735));// agility-pyramid
        // top(region12105)
        borders.addException(new ZoneBorders(3274, 3014, 3305, 3041));// Mining
        // camp
        borders.addException(new ZoneBorders(3260, 9408, 3331, 9472));// Mining
        // camp
        register(borders);
        pulse.stop();
    }

    @Override
    public boolean enter(Entity e) {
        if (e instanceof Player) {
            Player p = (Player) e;
            if (p.getAttribute("tutorial:stage", 0) > 71) {
                return true;
            }
            p.setAttribute("desert-delay", GameWorld.getTicks() + getDelay(p));
            PLAYERS.add(p);
            if (!pulse.isRunning()) {
                pulse.restart();
                pulse.start();
                GameWorld.getPulser().submit(pulse);
            }
        }
        return true;
    }

    @Override
    public boolean leave(Entity e, boolean logout) {
        if (e instanceof Player) {
            PLAYERS.remove(e);
            e.removeAttribute("desert-delay");
        }
        return super.leave(e, logout);
    }

}

File: Server/src/main/content/region/desert/handlers/DesertZone.kt
package content.region.desert.handlers

import core.api.consts.Animations
import core.api.consts.Items
import core.api.getRegionBorders
import core.cache.def.impl.ItemDefinition
import core.game.node.entity.Entity
import core.game.node.entity.combat.ImpactHandler.HitsplatType
import core.game.node.entity.player.Player
import core.game.node.item.Item
import core.game.system.task.Pulse
import core.game.world.GameWorld.Pulser
import core.game.world.GameWorld.ticks
import core.game.world.map.zone.MapZone
import core.game.world.map.zone.ZoneBorders
import core.game.world.map.zone.ZoneBuilder
import core.game.world.update.flag.context.Animation
import core.plugin.Initializable
import core.plugin.Plugin
import core.tools.RandomFunction

@Initializable
class DesertZone : MapZone(DESERT_ZONE, true), Plugin<Any?> {


    override fun newInstance(arg: Any?): Plugin<Any?> {
        ZoneBuilder.configure(this)
        return this
    }

    override fun fireEvent(identifier: String, vararg args: Any): Any? {
        return null
    }

    override fun configure() {
        register(getRegionBorders(12589))
        register(getRegionBorders(12590))
        register(getRegionBorders(12591))
        register(getRegionBorders(12843))
        register(getRegionBorders(12844))
        register(getRegionBorders(12845))
        register(getRegionBorders(12847))
        register(getRegionBorders(12848))
        register(getRegionBorders(13099))
        register(getRegionBorders(13100))
        register(getRegionBorders(13101))
        register(getRegionBorders(13102))
        register(getRegionBorders(13103))
        register(getRegionBorders(13355))
        register(getRegionBorders(13356))
        register(getRegionBorders(13357))
        register(getRegionBorders(13359))
        register(getRegionBorders(13360))
        register(getRegionBorders(13361))
        register(getRegionBorders(13611))
        register(getRegionBorders(13612))
        register(getRegionBorders(13613))
        register(getRegionBorders(13614))
        register(getRegionBorders(13615))
        register(getRegionBorders(13616))
        register(getRegionBorders(13617))
        register(getRegionBorders(13872))
        register(getRegionBorders(13873))
        register(ZoneBorders(3264, 3072, 3327, 3116))
        pulse.stop()
    }

    override fun enter(e: Entity): Boolean {
        if (e is Player) {
            val player = e
            if (player.getAttribute(TUTORIAL_STAGE, 0) > 71) {
                return true
            }
            player.setAttribute(DESERT_DELAY, ticks + getDelay(player))
            PLAYERS.add(player)
            if (!pulse.isRunning) {
                pulse.restart()
                pulse.start()
                Pulser.submit(pulse)
            }
        }
        return true
    }

    override fun leave(e: Entity, logout: Boolean): Boolean {
        if (e is Player) {
            PLAYERS.remove(e)
            e.removeAttribute(DESERT_DELAY)
        }
        return super.leave(e, logout)
    }

    companion object {
        private val WATER_SKINS = arrayOf(
            Item(Items.WATERSKIN4_1823),
            Item(Items.WATERSKIN3_1825),
            Item(Items.WATERSKIN2_1827),
            Item(Items.WATERSKIN1_1829)
        )
        private val VESSILS = arrayOf(
            intArrayOf(Items.JUG_OF_WATER_1937, Items.JUG_1935),
            intArrayOf(Items.BUCKET_OF_WATER_1929, Items.BUCKET_1925),
            intArrayOf(Items.BOWL_OF_WATER_1921, Items.BOWL_1923),
            intArrayOf(Items.VIAL_OF_WATER_227, Items.VIAL_229)
        )
        private val ANIMATION = Animation(Animations.EAT_OLD_829)
        private val PLAYERS: MutableList<Player> = ArrayList(20)
        private val DESERT_DELAY = "desert-delay"
        private val TUTORIAL_STAGE = "tutorial:stage"
        private val TUTORIAL_COMPLETE = "tutorial:complete"
        private val DESERT_ZONE = "Desert Zone"
        private val pulse: Pulse = object : Pulse(3) {
            override fun pulse(): Boolean {
                for (player in PLAYERS) {
                    if (!player.getAttribute(TUTORIAL_COMPLETE, false) || player.interfaceManager.isOpened || player.interfaceManager.hasChatbox() || player.locks.isMovementLocked) {
                        continue
                    }
                    if (player.getAttribute(DESERT_DELAY, -1) < ticks) {
                        effect(player)
                    }
                }
                return PLAYERS.isEmpty()
            }
        }

        private fun effect(player: Player) {
            player.setAttribute(DESERT_DELAY, ticks + getDelay(player))
            evaporate(player)
            if (drink(player)) {
                return
            }
            player.impactHandler.manualHit(
                player,
                RandomFunction.random(1, if (player.location.y < 2990) 12 else 8),
                HitsplatType.NORMAL
            )
            player.packetDispatch.sendMessage("You start dying of thirst while you're in the desert.")
        }

        fun evaporate(player: Player) {
            for (i in VESSILS.indices) {
                if (player.inventory.contains(VESSILS[i][0], 1)) {
                    if (player.inventory.remove(Item(VESSILS[i][0]))) {
                        player.inventory.add(Item(VESSILS[i][1]))
                        player.packetDispatch.sendMessage("The water in your " + ItemDefinition.forId(VESSILS[i][0]).name.lowercase().replace("of water", "").trim { it <= ' ' } + " evaporates in the desert heat.")
                    }
                }
            }
        }

        fun drink(player: Player): Boolean {
            for (i in WATER_SKINS) {
                if (player.inventory.containsItem(i) && player.inventory.remove(i)) {
                    player.inventory.add(Item(i.id + 2))
                    player.animate(ANIMATION)
                    player.packetDispatch.sendMessage("You take a drink of water.")
                    return true
                }
            }
            if (player.inventory.contains(Items.WATERSKIN0_1831, 1)) {
                player.packetDispatch.sendMessage("Perhaps you should fill up one of your empty waterskins.")
            } else {
                player.packetDispatch.sendMessage("You should get a waterskin for any travelling in the desert.")
            }
            return false
        }

        private fun getDelay(player: Player): Int {
            var delay = 116
            if (player.equipment.contains(Items.DESERT_SHIRT_1833, 1)) {
                delay += 17
            }
            if (player.equipment.contains(Items.DESERT_ROBE_1835, 1)) {
                delay += 17
            }
            if (player.equipment.contains(Items.DESERT_BOOTS_1837, 1)) {
                delay += 17
            }
            return delay
        }
    }

}
